src/env-inl.h

```diff
@@ -101,6 +101,20 @@ inline void Environment::AsyncHooks::set_enable_callbacks(uint32_t flag) {
   fields_[kEnableCallbacks] = flag;
 }
 
+inline Environment::AsyncCallbackScope::AsyncCallbackScope(Environment* env)
+    : env_(env) {
+  env_->async_callback_cntr_++;
+}
+
+inline Environment::AsyncCallbackScope::~AsyncCallbackScope() {
+  env_->async_callback_cntr_--;
+  CHECK_GE(env_->async_callback_cntr_, 0);
+}
+
+inline bool Environment::AsyncCallbackScope::in_tick() {
```

Comment:
`in_tick` is meant to indicate that we're currently "in" the state of processing the next "tick" queue. If it weren't for needing to process that queue then making `MakeCallback` reentrant would be much easier.

I don't have a code sample that can reproduce that, possibly @indutny might, but that call itself isn't recursive. It's that from JS we ask the HTTPParser for more header information. Which would mean MakeCallback would be called twice. While at the same time it's possible for the HTTPParser to initiate the JS stack.